secp256kfun/src/vendor/hash_to_curve.rs

```diff
@@ -0,0 +1,568 @@
+        let mut hasher = H::default();
+        let mut xor_result = [0u8; 32]; // Hash32 output size
+        for j in 0..B_IN_BYTES {
+            xor_result[j] = b_0[j] ^ b_i[j];
+        }
+        hasher.update(&xor_result[..B_IN_BYTES]);
+        hasher.update(&[i]);
+        hasher.update(&dst_prime[..actual_dst_prime_len]);
+        b_i = hasher.finalize_fixed();
+
+        let copy_len = core::cmp::min(B_IN_BYTES, len_in_bytes - offset);
+        uniform_bytes[offset..offset + copy_len].copy_from_slice(&b_i[..copy_len]);
+        offset += copy_len;
+    }
+
+    uniform_bytes
+}
+
+/// Hash to field element - always returns 2 elements for hash_to_curve
+fn hash_to_field<H: Hash32 + digest::Update + BlockSizeUser>(
+    msg: &[u8],
+    dst: &[u8],
+) -> [FieldElement; 2] {
+    let len_in_bytes = 2 * L;
+    let uniform_bytes = expand_message_xmd::<H>(msg, dst, len_in_bytes);
+
+    let mut output = [FieldElement::ZERO; 2];
+
+    // Split the uniform_bytes into two L-byte chunks
+    let (first_half, second_half) = uniform_bytes.split_at(L);
+    output[0] = from_okm(first_half.try_into().unwrap());
+    output[1] = from_okm(second_half.try_into().unwrap());
+
+    output
+}
+
+/// Convert output keying material to field element
+fn from_okm(data: &[u8; L]) -> FieldElement {
+    // Construct d0 from first 24 bytes
+    let d0 = FieldElement::from_bytes_unchecked(&[
+        0, 0, 0, 0, 0, 0, 0, 0, data[0], data[1], data[2], data[3], data[4], data[5], data[6],
+        data[7], data[8], data[9], data[10], data[11], data[12], data[13], data[14], data[15],
+        data[16], data[17], data[18], data[19], data[20], data[21], data[22], data[23],
+    ]);
+
+    // Construct d1 from next 24 bytes
+    let d1 = FieldElement::from_bytes_unchecked(&[
+        0, 0, 0, 0, 0, 0, 0, 0, data[24], data[25], data[26], data[27], data[28], data[29],
+        data[30], data[31], data[32], data[33], data[34], data[35], data[36], data[37], data[38],
+        data[39], data[40], data[41], data[42], data[43], data[44], data[45], data[46], data[47],
+    ]);
+
+    d0 * F_2_192 + d1
+}
+
+/// Map field element to curve using Simplified SWU
+fn map_to_curve(u: FieldElement) -> AffinePoint {
+    let (x, y) = sswu_map(u);
+    let (x_iso, y_iso) = isogeny_map(x, y);
+
+    AffinePoint::new(x_iso, y_iso)
+}
+
+/// Simplified SWU map
+fn sswu_map(u: FieldElement) -> (FieldElement, FieldElement) {
+    let tv1 = u.square();
+    let tv3 = Z * tv1;
+    let mut tv2 = tv3.square();
+    let mut xd = tv2 + tv3;
+    let x1n = MAP_B * (xd + FieldElement::ONE);
+    xd = (xd * MAP_A.negate(1)).normalize();
+
+    let tv = Z * MAP_A;
+    xd.conditional_assign(&tv, xd.is_zero());
+
+    tv2 = xd.square();
+    let gxd = tv2 * xd;
+    tv2 = tv2 * MAP_A;
+
+    let mut gx1 = x1n * (tv2 + x1n.square());
+    tv2 = gxd * MAP_B;
+    gx1 = gx1 + tv2;
+
+    let mut tv4 = gxd.square();
+    tv2 = gx1 * gxd;
+    tv4 = tv4 * tv2;
```

Comment:
I'm pretty sure you can do: `tv4 *= tv2`. There's a few of places where this operator could be used to simplify the code in this function.